File: nomad-samples/js/event-instance-scheduler/helpers/process-dates.js
const Lookup = (key, value) => ({
    key: key,
    value: value,
});

const DAY_OF_WEEK = {
    Sunday: Lookup("7", '3fc3b5ec-88be-41c9-be8f-5199735f3603'),
    Monday: Lookup("1", '16bebfa8-65cc-451e-9744-d09d6c761e4a'),
    Tuesday: Lookup("2", '02782764-f644-43fa-89cb-40cd3a3c3ece'),
    Wednesday: Lookup("3", '1ca7cf91-5460-479f-84e5-e02cd51ca3f9'),
    Thursday: Lookup("4", '2691d391-e1b1-43b6-97e2-5fc6b39479ef'),
    Friday: Lookup("5", 'b1897795-37a3-4a6e-a702-93643fe2ecab'),
    Saturday: Lookup("6", 'cf4c7688-417f-48d4-8b9d-fc6e6132d34e'),
};

export default function getCheckboxDates(DATE_CHECKBOX, startTime)
{
// goes through checkbox elements and 
    const DATE_LIST = [];

    for(let dateIdx = 0; dateIdx < DATE_CHECKBOX.length; ++dateIdx)
    {
        if (DATE_CHECKBOX[dateIdx].checked)
        {
            let dateCheckboxDay = DATE_CHECKBOX[dateIdx].value;
            let dateMap = {
                description: dateCheckboxDay,
                id: DAY_OF_WEEK[dateCheckboxDay].value,
                properties: {
                    abbreviation: dateCheckboxDay.substr(0, 3),
                    name: dateCheckboxDay,
                    sortOrder: DAY_OF_WEEK[dateCheckboxDay].key
                }
            };
            DATE_LIST.push(dateMap);
        }
    }

    /*  calculates the first day instance is going to be set to based of of first day of week
        choosen by user
    */ 
    const DATE = new Date();
    
    let startIdx = 1;
    let minimum = DAY_OF_WEEK[DATE_CHECKBOX[0].value].key - DATE.getDay();
    
    while (minimum < 0)
    {
        let diff = DAY_OF_WEEK[DATE_CHECKBOX[dateIdx].value].key - DATE.getDay();
        if (diff > 0 || (diff === 0 && parseInt(startTime.substr(0,2)) - DATE.getHours() >= 1))
        {
            minimum = diff;
        }
    }

    console.log(minimum);
    for(let dateIdx = startIdx; dateIdx < DATE_CHECKBOX.length; ++dateIdx)
    {
        if (DATE_CHECKBOX[dateIdx].checked)
        {
            let diff = DAY_OF_WEEK[DATE_CHECKBOX[dateIdx].value].key - DATE.getDay();
            if (diff < minimum)
            {
                if (diff === 0 && parseInt(startTime.substr(0,2)) - DATE.getHours() < 1)
                {
                    continue;
                }
                else
                {
                    minimum = diff;
                }
            }
        }
    }

    DATE.setDate(DATE.getDate() + minimum);

    return { startDate : DATE, dates: DATE_LIST }
}

File: nomad-samples/admin/py/event_instance_scheduler/helpers/process_dates.py
import datetime

DAY_OF_WEEK = {
    'Sunday': {"SORT_ORDER": 7, "DAY_ID": '3fc3b5ec-88be-41c9-be8f-5199735f3603'},
    'Monday': {"SORT_ORDER": 1, "DAY_ID": '16bebfa8-65cc-451e-9744-d09d6c761e4a'},
    'Tuesday': {"SORT_ORDER": 2, "DAY_ID": '02782764-f644-43fa-89cb-40cd3a3c3ece'},
    'Wednesday': {"SORT_ORDER": 3, "DAY_ID": '1ca7cf91-5460-479f-84e5-e02cd51ca3f9'},
    'Thursday': {"SORT_ORDER": 4, "DAY_ID": '2691d391-e1b1-43b6-97e2-5fc6b39479ef'},
    'Friday': {"SORT_ORDER": 5, "DAY_ID": 'b1897795-37a3-4a6e-a702-93643fe2ecab'},
    'Saturday': {"SORT_ORDER": 6, "DAY_ID": 'cf4c7688-417f-48d4-8b9d-fc6e6132d34e'}
}

def process_dates(DATES, START_TIME):
    SORTED_DATES = sorted(DATES, key=lambda day: list(DAY_OF_WEEK.keys()).index(day))

    DATE_LIST = []

    for DATE in SORTED_DATES:
        date_map = {
            "description": DATE,
            "id": DAY_OF_WEEK[DATE]["DAY_ID"],
            "properties": {
                "abbreviation": DATE[:3],
                "name": DATE,
                "sortOrder": str(DAY_OF_WEEK[DATE]["SORT_ORDER"])
            }
        }
        DATE_LIST.append(date_map)

    DATETIME = datetime.datetime.now()

    start_idx = 1
    minimum = DAY_OF_WEEK[SORTED_DATES[0]]["SORT_ORDER"] - 1 - DATETIME.weekday()
    while minimum < 0:
        diff = DAY_OF_WEEK[SORTED_DATES[start_idx]]["SORT_ORDER"] - 1 - DATETIME.weekday()
        if diff > 0 or (diff == 0 and int(START_TIME[:2]) - DATETIME.hour >= 1):
            minimum = diff

    for DATE_IDX in range(start_idx, len(SORTED_DATES)):
        diff = DAY_OF_WEEK[SORTED_DATES[DATE_IDX]]["SORT_ORDER"] - 1 - DATETIME.weekday()
        if diff < minimum:
            if diff == 0 and int(START_TIME[:2]) - DATETIME.hour < 1:
                continue
            else:
                minimum = diff

    FIRST_DATE = datetime.timedelta(days=minimum)
    FIRST_DATE += DATETIME
    return {"startDate": FIRST_DATE.strftime("%Y-%m-%d"), "dateList": DATE_LIST}
